### DISCUSSION TITLE
CoupledVarThresholdElementSubdomainModifier: x, y, z dependence

### DISCUSSION THREAD

**Discussion by czadeh** [idaholab/moose#20064] [Q&A Meshing] [answered]
Hi,

I have been trying to define a variable to be used with CoupledVarThresholdElementSubdomainModifier so that certain points on the mesh are modified while others are not. I encoded a spatial dependence on x, y, z points and made a class that inherits from Function.h. It returns a 0 or 1 to an AuxVariable if conditions are met. If the result is 1, then CoupledVarThresholdElementSubdomainModifier modifies the subdomain.

However when testing this function, the spatial dependence was nonexistent and it seemed that only the last value of the function was used for the entire mesh. 

I decided to try adapting the code to a new class that inherits from ADMaterial.h. This way each qp can be modified depending on the spatial conditions. However, when using the virtual function computeQpProperties(), I cannot easily use the x, y, z points like I could when using the virtual function value() in a class that inherits from Function.h.

I cannot use `const Point & p` as an argument to `computeQpProperties`. I tried to use `libMesh::Point(0)`, `libMesh::Point(1)`, `libMesh::Point(2)` as x, y, and z within my virtual function override but I am getting errors such as `'libMesh::TypeVector<T>::TypeVector(const T&, const T&, const T&) [with T = double]' is protected within this context`. 

I am not sure how to deal with this issue or whether I am taking the right approach. 

How can I use spatial data within computeQpProperties?

Is a material property output even compatible with CoupledVarThresholdElementSubdomainModifier?

Is there a better way to structure the output of a function so it has spatial dependence?

Thank you,

Cameron

**Comment by GiudGiud** [comment, marked as answer]
Hello

From a material computeQpProperties routine, you can access spatial properties using:
_q_point[_qp]

See this code for example:
https://github.com/idaholab/moose/blob/next/framework/src/materials/GenericFunctionMaterial.C

Guillaume

**Reply by czadeh** [reply]
Thank you, I also saw here how the authors declared x, y, and z starting from line 56. I used that convention.

```
const Real & x = _q_point[_qp](0);

const Real & y = _q_point[_qp](1);

const Real & z = _q_point[_qp](2);
```
I used my new material property along with a constant, monomial aux variable and the MaterialRealAux kernel to make it work with CoupledVarThresholdElementSubdomainModifier.
